compiler/optimizer/VPHandlers.cpp

```diff
@@ -12223,12 +12280,23 @@ TR::Node *constrainArrayStoreChk(OMR::ValuePropagation *vp, TR::Node *node)
       TR::VPConstraint *object = vp->getConstraint(objectRef, isGlobal);
       TR::VPConstraint *array  = vp->getConstraint(arrayRef, isGlobal);
 
-      // If the object reference is null we can remove this check.
+      const bool areValueTypesEnabled = TR::Compiler->om.areValueTypesEnabled();
+      TR_YesNoMaybe arrayCompIsValueType = isArrayCompTypeValueType(vp, array);
+
+      // If the object reference is null we can remove this check, if the
+      // array's element type is not a value type
+      // Null references are not allowed for value types, so only remove the
+      // check if the array type is resolved and cannot be of a type whose
+      // actual component type can be a value type
+      //
+      if (object && object->isNullObject() && arrayCompIsValueType == TR_no)
+         {
+         canBeRemoved = true;
+         }
+
       // If the array reference is null we can remove this check, since there
       // will be a nullcheck on the array reference before storing into it.
       //
-      if (object && object->isNullObject())
```

Comment:
can't this stay for when value types are not enabled? Save going all the way down and also having dependencies on constraints etc? That would minimize the delta for older versions.

Reply:
`isArrayCompTypeValueType` will always return `TR_no` if value types are not enabled, so the revised condition 
```
if (object && object->isNullObject() && arrayCompIsValueType == TR_no)
```
will catch the same cases as before when value types are not enabled.

Reply:
Thanks this looks less invasive